src/Octopurls.Tests/RedirectsURLsTests.cs

```diff
@@ -0,0 +1,140 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Octopurls.Tests
+{
+    public class RedirectsURLsTests : OctopurlTest
+    {
+        [Fact]
+        public void ParallelForeachTestRedirectsURLs()
+        {
+            var badURLs = new List<string>();
+
+            Parallel.ForEach(redirects.Urls, (url) =>
+            {
+                if (TestURL(url.Value) == false)
+                {
+                    badURLs.Add(url.Value);
+                }
+            });
+
+            if (badURLs.Any())
+            {
+                Console.WriteLine("The bad urls are:");
+                foreach (var badUrL in badURLs)
+                {
+                    Console.WriteLine($"{badUrL}");
+                }
+            }
+
+            Assert.Equal(0, badURLs.Count);
+        }
+
+        [Fact]
+        public void PLinqTestRedirectsURLs()
+        {
+
+            var badURLs = new List<string>();
+
+            var query = from url in redirects.Urls.AsParallel().AsOrdered()
+                where TestURL(url.Value) == false
+                select url.Value;
+
+            badURLs.AddRange(query.ToList());
```

Comment:
Looks nice! A really powerful thing about using LINQ: you don't need to create empty collections and fill them. It's fine in this scenario, but it breaks down in concurrent situations. I'd suggest in this case getting rid of `var badURLs = new list<string>();` and filling it later. Under the hood you are actually allocating the same list several times over.

Instead do `var badUrls = query.ToArray()` or `.ToList()` depending on what you want.

I would also suggest trying to get rid of `foreach` wherever it makes sense. In this case you could:

`var message = $"The bad urls are:{Environment.NewLine}{string.Join(Environment.NewLine, badUrls)}";`

Boom, one line of code inside that if statement.

Actually, one final thing, I wouldn't `Console.Writeline` the failure results. What you should do is take the formatted string with the bad urls and add it to the `Assert.AreEqual()` statement as the reason the test failed. It will get reported more clearly.

Reply:
That last bit probably came from me giving a poor example. `Assert.IsEmpty(badUrls)` would probably be better. That should emit the items if there are any, IIRC